File: data/scrape_skill_tree.py
#!/usr/bin/env python3
"""
Scrape d4planner.io build guides for all classes.
Each page shows a full skill tree with CategoryIcon row dividers
and MajorSkillIcon/MinorSkillIcon skill nodes.

Extracts icon IDs from background-image URLs, maps to skill names
via Maxroll JSON, and assigns rows using CategoryIcon Y positions
as dividers + known thresholds (0, 1, 2, 11, 16, 23, 33).
"""

import json
import re
import time
from pathlib import Path
from playwright.sync_api import sync_playwright

OUTPUT = Path(__file__).parent / "skill_tree_positions.json"
MAXROLL_JSON = (lambda: __import__("_maxroll").MAXROLL_PATH)()

# One build guide per class on d4planner
CLASS_URLS = {
    "Barbarian":   "https://d4planner.io/builds/barbarian-hammer-of-the-ancients-endgame-build-guide",
    "Druid":       "https://d4planner.io/builds/lightning-storm-druid-guide",
    "Necromancer": "https://d4planner.io/builds/bone-spear-necromancer-guide",
    "Rogue":       "https://d4planner.io/builds/rapid-fire-rogue-guide",
    "Sorcerer":    "https://d4planner.io/builds/ball-lightning-sorcerer-guide",
    "Spiritborn":  "https://d4planner.io/builds/lightning-evade-spiritborn-guide",
    # No Paladin guide on d4planner yet
}

# Confirmed in-game thresholds per row
ROW_THRESHOLDS = [0, 1, 2, 11, 16, 23, 33]


def build_icon_to_skill():
    with open(MAXROLL_JSON) as f:
        mdata = json.load(f)
    skills = mdata.get("skills") or {}
    if isinstance(skills, list):
        skills = {str(i): s for i, s in enumerate(skills)}
    mapping = {}
    for sid, s in skills.items():
        name = s.get("name") or ""
        if not name:
            continue
        for icon_id in (s.get("icons") or []):
            if icon_id and icon_id != 0:
                mapping[str(icon_id)] = name
    return mapping


def scrape_class(page, cls_name, url, icon_to_skill):
    print(f"  [{cls_name}] Loading {url}...")
    page.goto(url, wait_until="domcontentloaded", timeout=30000)
    time.sleep(4)

    # Dismiss cookie
    try:
        btn = page.query_selector("#onetrust-accept-btn-handler")
        if btn and btn.is_visible():
            btn.click()
            time.sleep(1)
    except Exception:
        pass

    # Extract all nodes with icon IDs and positions
    raw_nodes = page.evaluate("""() => {
        const tree = document.querySelector("[class^='Builds_miniSkillTreeContent']");
        if (!tree) return [];
        const nodes = tree.querySelectorAll("[class*='SkillNode_BaseNode']");
        const results = [];
        for (const node of nodes) {
            const top = parseFloat(node.style.top) || 0;
            const left = parseFloat(node.style.left) || 0;
            const classes = node.className || '';

            // Extract icon ID from background-image (computed style)
            let iconId = '';
            for (const child of [node, ...node.querySelectorAll('*')]) {
                const bg = window.getComputedStyle(child).backgroundImage || '';
                const m = bg.match(/(\\d{5,})\\.webp/);
                if (m) { iconId = m[1]; break; }
            }

            const isCategory = classes.includes('CategoryIcon');
            const isMajor = classes.includes('MajorSkillIcon');
            const isMinor = classes.includes('MinorSkillIcon');
            const isPassive = classes.includes('PassiveSkillIcon');
            const isKey = classes.includes('KeyPassive') || classes.includes('KeystoneIcon');

            results.push({
                top, left, iconId, isCategory, isMajor, isMinor, isPassive, isKey,
                classes: classes.substring(0, 100),
            });
        }
        return results;
    }""")

    print(f"  [{cls_name}] {len(raw_nodes)} nodes extracted")

    # Separate category icons (row headers) from skill nodes
    categories = sorted([n for n in raw_nodes if n["isCategory"]], key=lambda n: n["top"])
    skills = [n for n in raw_nodes if not n["isCategory"]]

    print(f"  [{cls_name}] {len(categories)} category dividers, {len(skills)} skill nodes")

    # Category icons define row boundaries
    # There should be 5-7 of them (one per row: Basic, Core, Row3, Row4, Row5, Ultimate, Key)
    # Map each to a threshold
    if len(categories) >= 5:
        # Sort by Y, assign thresholds in order
        for i, cat in enumerate(categories):
            threshold = ROW_THRESHOLDS[i] if i < len(ROW_THRESHOLDS) else 33
            cat["threshold"] = threshold
            name = icon_to_skill.get(cat["iconId"], f"Row{i+1}")
            print(f"    Category {i+1}: Y={cat['top']:.0f} pts={threshold} icon={cat['iconId']} name={name}")
    else:
        print(f"    WARNING: expected 5+ categories, got {len(categories)}")

    # Assign each skill to a row based on Y position
    results = {}
    for n in skills:
        icon_id = n.get("iconId", "")
        name = icon_to_skill.get(icon_id)
        if not name:
            continue

        y = n["top"]

        # Find which category this node falls under
        pts = 0
        for cat in categories:
            if y >= cat["top"] - 30:  # tolerance for nodes slightly above header
                pts = cat.get("threshold", 0)

        # Key passives override
        if n.get("isKey"):
            pts = 33

        key = name.lower()
        # Don't overwrite if already found at a different position
        # (some skills appear multiple times as enhanced/morph variants)
        if key not in results:
            results[key] = {
                "name": name,
                "y": round(y),
                "points_required": pts,
                "icon_id": icon_id,
            }

    print(f"  [{cls_name}] {len(results)} unique skills mapped")
    return results


def main():
    icon_to_skill = build_icon_to_skill()
    print(f"Icon mapping: {len(icon_to_skill)} icons\n")

    all_results = {}

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=False)
        page = browser.new_page()

        for cls_name, url in CLASS_URLS.items():
            try:
                results = scrape_class(page, cls_name, url, icon_to_skill)
                all_results[cls_name] = results
                print()
            except Exception as e:
                print(f"  [{cls_name}] ERROR: {e}\n")
                all_results[cls_name] = {}

        browser.close()

    with open(OUTPUT, "w") as f:
        json.dump(all_results, f, indent=2, sort_keys=True)

    print(f"\nWrote {OUTPUT}")
    for cls, skills in all_results.items():
        print(f"  {cls}: {len(skills)} skills")


if __name__ == "__main__":
    main()

File: data/scrape_skill_tree.js
#!/usr/bin/env node
/**
 * Scrape Maxroll D4 skill tree planner to extract every skill/passive
 * and its row position. Uses Playwright to render the page and hover
 * over each node to get tooltip data.
 *
 * Output: data/skill_tree_positions.json
 *   { "class": { "skill_name": { row, section, points_required } } }
 */

import { chromium } from 'playwright';
import { writeFileSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';

const __dirname = dirname(fileURLToPath(import.meta.url));

// Maxroll skill tree planner URLs per class
const CLASS_URLS = {
  Barbarian:   'https://maxroll.gg/d4/planner#c=barbarian',
  Druid:       'https://maxroll.gg/d4/planner#c=druid',
  Necromancer: 'https://maxroll.gg/d4/planner#c=necromancer',
  Rogue:       'https://maxroll.gg/d4/planner#c=rogue',
  Sorcerer:    'https://maxroll.gg/d4/planner#c=sorcerer',
  Spiritborn:  'https://maxroll.gg/d4/planner#c=spiritborn',
  Paladin:     'https://maxroll.gg/d4/planner#c=paladin',
};

// In-game point thresholds per row (1-indexed)
const ROW_POINTS = { 1: 0, 2: 1, 3: 2, 4: 11, 5: 16, 6: 23, 7: 33 };

async function scrapeClass(browser, className, url) {
  const page = await browser.newPage();
  console.log(`  [${className}] Loading ${url}...`);

  await page.goto(url, { waitUntil: 'networkidle', timeout: 30000 });

  // Wait for skill tree to render
  await page.waitForSelector('.d4t-SkillPreview, .d4t-TreeGraph', { timeout: 15000 }).catch(() => {
    console.log(`  [${className}] No skill tree found, trying expanded view...`);
  });

  // Try to find the condensed view (d4t-SkillPreview with d4t-skill-group rows)
  const hasCondensed = await page.$('.d4t-SkillPreview .d4t-skill-group');

  let results = {};

  if (hasCondensed) {
    console.log(`  [${className}] Using condensed view...`);
    results = await scrapeCondensedView(page, className);
  } else {
    // Try expanded tree view
    console.log(`  [${className}] Using expanded view, hovering nodes...`);
    results = await scrapeExpandedView(page, className);
  }

  await page.close();
  return results;
}

async function scrapeCondensedView(page, className) {
  const results = {};

  // Get all skill groups (rows)
  const groups = await page.$$('.d4t-skill-group');
  console.log(`  [${className}] Found ${groups.length} rows`);

  for (let rowIdx = 0; rowIdx < groups.length; rowIdx++) {
    const row = rowIdx + 1;
    const pts = ROW_POINTS[row] ?? 16;
    const group = groups[rowIdx];

    // Get all nodes in this row — hover each to get the tooltip
    const nodes = await group.$$('.d4t-skill-frame, .d4t-passive-frame');

    for (const node of nodes) {
      try {
        // Hover to trigger tooltip
        await node.hover({ timeout: 2000 });
        await page.waitForTimeout(150);

        // Read tooltip
        const tooltip = await page.$('.d4t-Tooltip, .d4-tooltip, [class*="tooltip"]');
        if (tooltip) {
          const name = await tooltip.$eval(
            '.d4t-skill-name, .d4t-header, [class*="name"]',
            el => el.textContent?.trim()
          ).catch(() => null);

          if (name && name.length > 1 && !name.startsWith('?')) {
            results[name.toLowerCase()] = {
              name,
              row,
              points_required: pts,
            };
          }
        }
      } catch (e) {
        // Skip nodes that can't be hovered
      }
    }

    console.log(`  [${className}] Row ${row} (${pts}pts): ${Object.keys(results).length} skills so far`);
  }

  return results;
}

async function scrapeExpandedView(page, className) {
  const results = {};

  // In expanded view, nodes have absolute positions
  // Hover each node to get tooltip with skill name
  const nodes = await page.$$('.d4t-node');
  console.log(`  [${className}] Found ${nodes.length} nodes in expanded view`);

  for (let i = 0; i < nodes.length; i++) {
    try {
      const node = nodes[i];
      await node.hover({ timeout: 2000 });
      await page.waitForTimeout(150);

      // Try to read tooltip
      const tooltipText = await page.$eval(
        '.d4t-Tooltip .d4t-skill-name, .d4t-Tooltip [class*="name"]',
        el => el.textContent?.trim()
      ).catch(() => null);

      if (tooltipText && tooltipText.length > 1) {
        // Get Y position for row estimation
        const box = await node.boundingBox();
        if (box) {
          results[tooltipText.toLowerCase()] = {
            name: tooltipText,
            y: Math.round(box.y),
          };
        }
      }
    } catch (e) {
      // Skip
    }

    if (i % 20 === 0) {
      console.log(`  [${className}] Processed ${i}/${nodes.length} nodes...`);
    }
  }

  return results;
}

async function main() {
  console.log('Launching browser...');
  const browser = await chromium.launch({ headless: true });

  const allResults = {};

  for (const [className, url] of Object.entries(CLASS_URLS)) {
    try {
      allResults[className] = await scrapeClass(browser, className, url);
      console.log(`  [${className}] Done: ${Object.keys(allResults[className]).length} skills\n`);
    } catch (e) {
      console.log(`  [${className}] ERROR: ${e.message}\n`);
      allResults[className] = {};
    }
  }

  await browser.close();

  const outPath = join(__dirname, 'skill_tree_positions.json');
  writeFileSync(outPath, JSON.stringify(allResults, null, 2));
  console.log(`\nWrote ${outPath}`);

  // Summary
  for (const [cls, skills] of Object.entries(allResults)) {
    console.log(`  ${cls}: ${Object.keys(skills).length} skills`);
  }
}

main().catch(console.error);
